### PR TITLE
[Wasm EH] Optimize throws caught by TryTable into breaks

### PR DESCRIPTION
**Pull request by kripken**
E.g.

```wat
(try_table (catch_all $catch)
  (throw $e)
)

=>

(try_table (catch_all $catch)
  (br $catch)
)
```

This can then allow other passes to remove the TryTable, if no throwing
things remain.

### CODE REVIEW

**Review by aheejin — COMMENTED** [review]
LGTM! By the way why would this routine be in RemoveUnusedBr, given that it rather creates branches? Not that I have another place in mind.. (Other than let's say Vacuum which has become a kitchensync of sorts...)

**Comment by kripken** [comment]
> By the way why would this routine be in RemoveUnusedBr, given that it rather creates branches?

My reasoning is that by lowering a throw to a break we can then optimize that break right here in the pass, without waiting to reach another pass (e.g. a test here shows a throw turning into a br that turn turns into a br_if).

But, yeah, this could have fit elsewhere. I see Vacuum more as "remove obviously unneeded (but not dead) code", so if I had to pick another place I'd use OptimizeInstructions, though this is a little more complex than the typical peephole optimization there...

File: src/passes/RemoveUnusedBrs.cpp [hunk 1]
```diff
@@ -260,7 +261,7 @@ struct RemoveUnusedBrs : public WalkerPass<PostWalker<RemoveUnusedBrs>> {
     } else if (curr->is<Nop>()) {
       // ignore (could be result of a previous cycle)
       self->stopValueFlow();
-    } else if (curr->is<Loop>()) {
+    } else if (curr->is<Loop>()) { // TODO: eh
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/RemoveUnusedBrs.cpp:264]: Just that we can optimize exception handling here, both Try and TryTable. I'll do that soon.

[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/RemoveUnusedBrs.cpp:264]: (the handling will be similar to Loop)

**Comment by aheejin** [comment]
If a branch is created here, can it be optimized away in the same run of the pass? Apparently it does, as you've shown with the `br_if` example... Does that mean it traverses the program more than once?

**Comment by kripken** [comment]
Yes, this does loop over several simplifications, and keeps going until things stop changing. That main loop starts here:

https://github.com/WebAssembly/binaryen/blob/7e1413902d74111898419b9ed5add45429fa62f8/src/passes/RemoveUnusedBrs.cpp#L920-L923

This could probably be more efficient, but each traversal is practically linear. This pass didn't show up as slow in any of my recent performance measurements.

**Review by aheejin — APPROVED** [review]
I see, then it makes sense to put it here. Thanks!

**Pull request merged**